nemo/tron/converter/common.py

```diff
@@ -0,0 +1,346 @@
+
+
+@contextmanager
+def temporary_distributed_context():
+    if "MASTER_ADDR" not in os.environ:
+        os.environ["MASTER_ADDR"] = "localhost"
+    if "MASTER_PORT" not in os.environ:
+        os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group(backend="gloo", world_size=1, rank=0)
+    parallel_state.initialize_model_parallel()
+    yield
+    parallel_state.destroy_model_parallel()
+    dist.destroy_process_group()
+
+
+def get_full_mcore_state_dict(dist_ckpt_folder: Path, model_cfg):
+    with temporary_distributed_context():
+        if model_cfg.params_dtype != torch_dtype_from_mcore_config(model_cfg):
+            logger.info(
+                f"Converting params_dtype from {model_cfg.params_dtype} to {torch_dtype_from_mcore_config(model_cfg)}"
+            )
+            model_cfg.params_dtype = torch_dtype_from_mcore_config(model_cfg)
+
+        with megatron_cpu_init_context(model_cfg):
+            model = model_cfg.configure_model(None)
+
+        strategy = TorchDistLoadShardedStrategy()
+        state_dict = strategy.load(model.sharded_state_dict(), Path(dist_ckpt_folder))
+        del model
+
+    return state_dict
+
+
+def save_hf_tokenizer_assets(tokenizer_name_or_path, save_path="/tmp/nemo_tokenizer"):
+    """Save HF tokenizer to the imported NeMo model"""
+    from transformers import AutoTokenizer
+
+    tok = AutoTokenizer.from_pretrained(tokenizer_name_or_path, trust_remote_code=True)
+    # Save tokenizer assets to save_path.
+    tok.save_pretrained(save_path)
+    return save_path
+
+
+def dtype_from_str(dtype):
+    """
+    Convert a str precision to equivalent torch dtype.
+    """
+    assert isinstance(dtype, str)
+    if dtype in ["float16", "fp16", "16", "16-mixed"]:
+        return torch.float16
+    elif dtype in ["bfloat16", "bf16-mixed"]:
+        return torch.bfloat16
+    else:
+        return torch.float32
+
+
+def dtype_from_hf(config):
+    """
+    Extracts torch dtype from a HF config
+    """
+    assert hasattr(config, "torch_dtype"), "Expected config to have attr `torch_dtype`"
+    torch_dtype = config.torch_dtype
+    if isinstance(torch_dtype, torch.dtype):
+        return torch_dtype
+    elif isinstance(torch_dtype, str):
+        return dtype_from_str(torch_dtype)
+    else:
+        raise ValueError("torch_dtype is not of type str/torch.dtype")
+
+
+class _ModelState:
+    """
+    Helper class for used for to modify state dict of a source model during model conversion.
+    """
+
+    def __init__(self, state_dict):
+        self._state_dict = state_dict
+
+    def state_dict(self):
+        # pylint: disable=C0115,C0116
+        return self._state_dict
+
+    def to(self, dtype):
+        # pylint: disable=C0115,C0116
+        for k, v in self._state_dict.items():
+            if v.dtype != dtype:
+                logger.warning(f"Converting {k} from {v.dtype} (source model) to {dtype} (target model)")
+            self._state_dict[k] = v.to(dtype)
+
+
+class BaseImporter:
```

Comment:
nit: should this be an abstract base class for the unimplemented methods?

Reply:
Yeah I can make this an abstract base class.